### PR TITLE
Fix LensFlare occlusion on scene view

### PR DESCRIPTION
**Pull request by skhiat**
Fix flickering of LensFlare Occlusion with TAA (so HDRP only).
By doing accumulation "unanimity vote" through multiple frame (8 by default)
Before this PR:
1LGEzSRXeS
With the PR:
Inszu7cA68

QA:
- Flickering occurs before when the LensFlare had occlusion enable and place very close to an edge with TAA enabled
- Should extended testing with various parameter of TAA to see if the flickering is fixed
- Max-Lens-Flare-With-Occlusion: supported 128. Variables available in the code LensFlareCommonSRP.maxLensFlareWithOcclusion;

### CODE REVIEW

**Comment by sebastienlagarde** [comment]
Hey, what's the status of this PR?

**Review by iM0ve — COMMENTED** [review]
Hey, thanks for adding the showcase of before/after. Was there something else tested or needs to be tested? Any tips what QA should look at in more detail?

**Comment by skhiat** [comment]
> Hey, thanks for adding the showcase of before/after. Was there something else tested or needs to be tested? Any tips what QA should look at in more detail?

Updated.

File: com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/LensFlareMergeOcclusionDataDriven.compute [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#ifdef DEBUG_DISPLAY
+#include "Packages/com.unity.render-pipelines.core/ShaderLibrary/Debug.hlsl"
+#include "Packages/com.unity.render-pipelines.high-definition/Runtime/Debug/DebugDisplay.hlsl"
+#endif
+
+RW_TEXTURE2D_X(float, _LensFlareOcclusion);
+
+[numthreads(8, 1, 1)]
+void MainCS(uint3 dispatchThreadId : SV_DispatchThreadID)
```
Comment (kecho) [inline com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/LensFlareMergeOcclusionDataDriven.compute:30]: @soufianekhiat while this is fine for now, I wanna challenge you:
You have an entire wave (64 threads) that you could exploit, you could dispatch a 8x8 (64 threads) and in the y coordinate load your history into lds.
You can then do a parallel reduction or interlock min on a final value and avoid tapping 8 times the history resource :)

Just fun programming if you are up to it! Please dont consider this as a blocker

Comment (skhiat) [reply com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/LensFlareMergeOcclusionDataDriven.compute]: Updated with 3 barriers & LDS

**Review by iM0ve — APPROVED** [review]
Looks good. I reproduced the issue on master and then tested:
* Issue no longer appears with default settings in Scene and Game views
* Modified all TAA settings in an attempt to reintroduce flickering
* Used all Lens flare occlusion settings in an attempt to reintroduce flickering
* Tested with several lights visible on screen
* Tested occlusion that is based on offscreen instead of geometry

https://user-images.githubusercontent.com/36502659/134671045-6a6a759c-1819-4605-a25f-54cd4d184915.mp4


https://user-images.githubusercontent.com/36502659/134671059-2daf45ea-aa69-4edb-8e4c-cf28af5f9211.mp4

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs [hunk 2]
```diff
@@ -313,6 +337,182 @@ static Vector3 WorldToViewportDistance(Camera cam, Vector3 positionWS)
...
+        /// <param name="cam">Camera</param>
+        /// <param name="actualWidth">Width actually used for rendering after dynamic resolution and XR is applied.</param>
+        /// <param name="actualHeight">Height actually used for rendering after dynamic resolution and XR is applied.</param>
+        /// <param name="usePanini">Set if use Panani Projection</param>
+        /// <param name="paniniDistance">Distance used for Panini projection</param>
+        /// <param name="paniniCropToFit">CropToFit parameter used for Panini projection</param>
+        /// <param name="viewProjMatrix">View Projection Matrix of the current camera</param>
+        /// <param name="cmd">Command Buffer</param>
+        /// <param name="colorBuffer">Source Render Target which contains the Color Buffer</param>
+        /// <param name="_FlareTex">ShaderID for the FlareTex</param>
```
Comment (kecho) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs:354]: The parameters dont seam to match the documentation.

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs [hunk 3]
```diff
@@ -339,11 +539,12 @@ static Vector3 WorldToViewportDistance(Camera cam, Vector3 positionWS)
         static public void DoLensFlareDataDrivenCommon(Material lensFlareShader, LensFlareCommonSRP lensFlares, Camera cam, float actualWidth, float actualHeight,
             bool usePanini, float paniniDistance, float paniniCropToFit,
             bool isCameraRelative,
+            Vector3 cameraPositionWS,
```
Comment (kecho) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs:542]: Missing docs?

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs [hunk 4]
```diff
@@ -8,11 +8,35 @@ public sealed class LensFlareCommonSRP
...
+        static public void Initialize()
+        {
+            if (occlusionRT == null)
+                occlusionRT = RTHandles.Alloc(width: maxLensFlareWithOcclusion, height: maxLensFlareWithOcclusionTemporalSample + 1 * mergeNeeded, colorFormat: Experimental.Rendering.GraphicsFormat.R16_SFloat, enableRandomWrite: true, dimension: TextureXR.dimension);
+        }
+
+        static public void Dispose()
+        {
+            if (occlusionRT != null)
+                RTHandles.Release(occlusionRT);
```
Comment (kecho) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs:37]: Set to null after Releasing.

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs [hunk 5]
```diff
@@ -313,6 +337,182 @@ static Vector3 WorldToViewportDistance(Camera cam, Vector3 positionWS)
...
+
+                Vector2 screenPos = new Vector2(2.0f * viewportPos.x - 1.0f, 1.0f - 2.0f * viewportPos.y);
+
+                Vector2 radPos = new Vector2(Mathf.Abs(screenPos.x), Mathf.Abs(screenPos.y));
+                float radius = Mathf.Max(radPos.x, radPos.y); // l1 norm (instead of l2 norm)
+                float radialsScaleRadius = comp.radialScreenAttenuationCurve.length > 0 ? comp.radialScreenAttenuationCurve.Evaluate(radius) : 1.0f;
+
+                float currentIntensity = comp.intensity * radialsScaleRadius * distanceAttenuation;
+
+                if (currentIntensity <= 0.0f)
```
Comment (kecho) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs:486]: Perf, can we move this earlier in the loop?

**Review by kecho — CHANGES_REQUESTED** [review]
Requesting some minor updates. Please see comments.

**Review by kecho — APPROVED** [review]

**Pull request merged**